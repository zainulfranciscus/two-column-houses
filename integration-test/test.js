import { Selector } from "testcafe";

fixture`Adding houses into the right column`.page`http://localhost:3000/`;

const house = Selector("#container-1");
const addBtn = "#move-btn-1";
const houseNumber1InTheRightColumn = Selector("#right-column #container-1");

test("Clicking on an add button adds a house to the right column", async t => {

  await t
    .hover(house)
    .click(addBtn)
    .expect(houseNumber1InTheRightColumn.count)
    .eql(1);
});

test("Clicking on an add button twice, will only add a house once to the right column", async t => {

  await t
    .hover(house)
    .click(addBtn)
    .click(addBtn)
    .expect(houseNumber1InTheRightColumn.count)
    .eql(1);
});

test("Clicking on a remove button, removes a house in the right column", async t => {
  const house = Selector("#container-4");
  const removeBtn = "#move-btn-4";
  const houseInRightColumn = Selector("#right-column #container-4");
  await t
    .hover(house)
    .click(removeBtn)
    .expect(houseInRightColumn.count)
    .eql(0);
});


