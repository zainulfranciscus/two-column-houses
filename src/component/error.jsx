import React, { Component } from "react";

export const ERROR_MESSAGE_ID = "error";
export const ERROR_MESSAGE = "Unable to fetch JSON file from the given URL";

export default class Error extends Component {
  render() {
    if (!this.props.hasErrored) {
      return null;
    }

    return (
      <div id={ERROR_MESSAGE_ID} className="row error-container">
        <div className="columns small-12">{ERROR_MESSAGE}</div>
      </div>
    );
  }
}
