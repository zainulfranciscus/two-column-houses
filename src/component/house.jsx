import React, { Component } from "react";
import houseDecorator, { ATTRIBUTE_IS_MISSING } from "../data/house.decorator";

export const PREFIX_FOR_IMG_ID = "main-image-";
export const PREFIX_FOR_LOGO_ID = "logo-";
export const PREFIX_FOR_PRICE_ID = "price-";

export const PREFIX_FOR_LOGO_CONTAINER_ID = "logo-container-";
export const PREFIX_FOR_HOUSE_CONTAINER = "container-";
export const PREFIX_FOR_MOVE_BTN_ID = "move-btn-";
export const CSS_CLASS_FOR_HIDING_BUTTON = "is-hidden";
export const CSS_CLASS_FOR_BUTTON = "button";
export const PRICE_HAS_NOT_BEEN_SPECIFIED = "Price has not been set";

export function generateID(prefix, id) {
  return prefix + id;
}

export default class House extends Component {
  constructor(props) {
    super(props);

    this.state = {
      showButton: false
    };
  }

  moveHouseToAColumn() {
    const house = this.props.data;
    this.props.moveHouse(house);
  }

  displayButton() {
    this.setState({
      showButton: true
    });
  }

  hideButton() {
    this.setState({
      showButton: false
    });
  }

  render() {
    const decoratedHouseData = houseDecorator(this.props.data);
    const mainImageURL = decoratedHouseData.mainImage();
    const price = decoratedHouseData.price();
    const id = decoratedHouseData.id();
    const buttonText = this.props.buttonText;
    const logoURL = decoratedHouseData.logo();
    const brandingColor = decoratedHouseData.primaryBrandingColor();

    if (id === ATTRIBUTE_IS_MISSING) {
      return null;
    }

    return (
      <div
        id={generateID(PREFIX_FOR_HOUSE_CONTAINER, id)}
        className="property-container"
        onMouseOver={this.displayButton.bind(this)}
        onMouseLeave={this.hideButton.bind(this)}
      >
        {logoURL !== ATTRIBUTE_IS_MISSING && (
          <div
            id={generateID(PREFIX_FOR_LOGO_CONTAINER_ID, id)}
            style={{ backgroundColor: brandingColor }}
          >
            <img
              id={generateID(PREFIX_FOR_LOGO_ID, id)}
              src={logoURL}
              alt="agency logo"
            />
          </div>
        )}

        {mainImageURL !== ATTRIBUTE_IS_MISSING && (
          <img
            id={generateID(PREFIX_FOR_IMG_ID, id)}
            src={mainImageURL}
            alt="house"
          />
        )}

        <span id={generateID(PREFIX_FOR_PRICE_ID, id)}>
          {price === ATTRIBUTE_IS_MISSING
            ? PRICE_HAS_NOT_BEEN_SPECIFIED
            : price}
        </span>
        <div className="button-container">
          <button
            className={
              this.state.showButton
                ? CSS_CLASS_FOR_BUTTON
                : CSS_CLASS_FOR_BUTTON + " " + CSS_CLASS_FOR_HIDING_BUTTON
            }
            id={generateID(PREFIX_FOR_MOVE_BTN_ID, id)}
            onClick={this.moveHouseToAColumn.bind(this)}
          >
            {buttonText}
          </button>
        </div>
      </div>
    );
  }
}
