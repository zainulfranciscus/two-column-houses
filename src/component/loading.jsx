import React, { Component } from "react";

export const ID_OF_LOADER = "loader";
export const MESSAGE_WHILE_LOADING = "Please wait while I fetch house data";

export default class Loading extends Component {
  render() {
    if (!this.props.fetchingData || this.props.hasErrored) {
      return null;
    }

    return (
      <div className="row loading-component" id={ID_OF_LOADER}>
        <div className="columns small-12">{MESSAGE_WHILE_LOADING}</div>
      </div>
    );
  }
}
