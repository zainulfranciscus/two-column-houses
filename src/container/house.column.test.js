import React from "react";
import { HouseColumn, LEFT_COLUMN_ID, RIGHT_COLUMN_ID } from "./house.column";
import { findButton } from "../component/house.test";
import { aHouse } from "../data/house.decorator.test";
import { PREFIX_FOR_HOUSE_CONTAINER } from "../component/house";
import { mount } from "enzyme";
import {
  LEFT_COLUMN_HOUSES_PROPERTY_NAME,
  RIGHT_COLUMN_HOUSES_PROPERTY_NAME
} from "../data/list.of.houses.decorator";

describe("Unit tests for HouseColumn container ", () => {
  const leftColumnHousesOnly = {};
  leftColumnHousesOnly[LEFT_COLUMN_HOUSES_PROPERTY_NAME] = [aHouse];

  const rightColumnHousesOnly = {};
  rightColumnHousesOnly[RIGHT_COLUMN_HOUSES_PROPERTY_NAME] = [aHouse];

  it("when the container is given an empty list of houses, then no houses should be rendered", () => {
    const wrapper = mount(<HouseColumn houses={null} />);

    expect(housesInTheLeftColumn(wrapper).length).toBe(0);
    expect(housesInTheRightColumn(wrapper).length).toBe(0);
  });

  it("when the container is given an empty object of houses, then no houses should be rendered", () => {
    const wrapper = mount(<HouseColumn houses={{}} />);

    expect(housesInTheLeftColumn(wrapper).length).toBe(0);
    expect(housesInTheRightColumn(wrapper).length).toBe(0);
  });

  it("when the container is given a house into the left column, then 1 house should be rendered in the left column", () => {
    const wrapper = mount(<HouseColumn houses={leftColumnHousesOnly} />);

    expect(housesInTheLeftColumn(wrapper).length).toBe(1);
    expect(housesInTheRightColumn(wrapper).length).toBe(0);
  });

  it("when the container is given a house into the right column, then 1 house should be rendered in the right column", () => {
    const wrapper = mount(<HouseColumn houses={rightColumnHousesOnly} />);

    expect(housesInTheLeftColumn(wrapper).length).toBe(0);
    expect(housesInTheRightColumn(wrapper).length).toBe(1);
  });

  it("mock fetchListOfHouses should be called when component is mounted", () => {
    let mockFetchListOfHouses = jest.fn();
    mount(<HouseColumn fetchListOfHouses={mockFetchListOfHouses} />);
    expect(mockFetchListOfHouses.mock.calls.length).toBe(1);
  });

  it("when button is clicked then, the moveHouseToTheRightColumn is called ", () => {
    let moveHouseToTheRightColumn = jest.fn();
    const wrapper = mount(
      <HouseColumn
        houses={leftColumnHousesOnly}
        moveHouseToTheRightColumn={moveHouseToTheRightColumn}
      />
    );
    findButton(wrapper).simulate("click");
    expect(moveHouseToTheRightColumn.mock.calls.length).toBe(1);
  });

  it("when button is clicked then, the moveHouseToTheLeftColumn is called ", () => {
    let moveHouseToTheLeftColumn = jest.fn();
    const wrapper = mount(
      <HouseColumn
        houses={rightColumnHousesOnly}
        moveHouseToTheLeftColumn={moveHouseToTheLeftColumn}
      />
    );
    findButton(wrapper).simulate("click");
    expect(moveHouseToTheLeftColumn.mock.calls.length).toBe(1);
  });

  function housesInTheLeftColumn(wrapper) {
    const SELECTOR_FOR_CONTAINER_OF_LEFT_COLUMN_HOUSES =
      "div#" +
      LEFT_COLUMN_ID +
      " div[id^='" +
      PREFIX_FOR_HOUSE_CONTAINER +
      "']";

    return wrapper.find(SELECTOR_FOR_CONTAINER_OF_LEFT_COLUMN_HOUSES);
  }

  function housesInTheRightColumn(wrapper) {
    const SELECTOR_FOR_CONTAINER_OF_RIGHT_COLUMN_HOUSES =
      "div#" +
      RIGHT_COLUMN_ID +
      " div[id^='" +
      PREFIX_FOR_HOUSE_CONTAINER +
      "']";

    return wrapper.find(SELECTOR_FOR_CONTAINER_OF_RIGHT_COLUMN_HOUSES);
  }
});
