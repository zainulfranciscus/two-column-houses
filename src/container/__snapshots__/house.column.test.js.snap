// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`house component unit tests house component should match the most up to date snapshot 1`] = `
<div
  className="property-container"
  id="container-1"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <div
    id="logo-container-1"
    style={
      Object {
        "backgroundColor": "#ffe512",
      }
    }
  >
    <img
      alt="agency logo"
      id="logo-1"
      src="http://i1.au.reastatic.net/agencylogo/XRWXMT/12/20120927204448.gif"
    />
  </div>
  <img
    alt="house"
    id="main-image-1"
    src="http://i2.au.reastatic.net/640x480/20bfc8668a30e8cabf045a1cd54814a9042fc715a8be683ba196898333d68cec/main.jpg"
  />
  <span
    id="price-1"
  >
    $726,500
  </span>
  <button
    className="button is-hidden"
    id="move-btn-1"
    onClick={[Function]}
  >
    add
  </button>
</div>
`;
